src/scheduler.cpp

```diff
@@ -20,116 +20,69 @@
 #include "otpch.h"
 
 #include "scheduler.h"
+#include <boost/asio/post.hpp>
+#include <memory>
 
 void Scheduler::threadMain()
 {
-	std::unique_lock<std::mutex> eventLockUnique(eventLock, std::defer_lock);
-	while (getState() != THREAD_STATE_TERMINATED) {
-		std::cv_status ret = std::cv_status::no_timeout;
-
-		eventLockUnique.lock();
-		if (eventList.empty()) {
-			eventSignal.wait(eventLockUnique);
-		} else {
-			ret = eventSignal.wait_until(eventLockUnique, eventList.top()->getCycle());
-		}
-
-		// the mutex is locked again now...
-		if (ret == std::cv_status::timeout && !eventList.empty()) {
-			// ok we had a timeout, so there has to be an event we have to execute...
-			SchedulerTask* task = eventList.top();
-			eventList.pop();
-
-			// check if the event was stopped
-			auto it = eventIds.find(task->getEventId());
-			if (it == eventIds.end()) {
-				eventLockUnique.unlock();
-				delete task;
-				continue;
-			}
-			eventIds.erase(it);
-			eventLockUnique.unlock();
-
-			task->setDontExpire();
-			g_dispatcher.addTask(task, true);
-		} else {
-			eventLockUnique.unlock();
-		}
-	}
+	io_context.run();
 }
 
 uint32_t Scheduler::addEvent(SchedulerTask* task)
 {
-	eventLock.lock();
-
-	if (getState() != THREAD_STATE_RUNNING) {
-		eventLock.unlock();
-		delete task;
-		return 0;
-	}
-
 	// check if the event has a valid id
 	if (task->getEventId() == 0) {
-		// if not generate one
-		if (++lastEventId == 0) {
-			lastEventId = 1;
-		}
-
-		task->setEventId(lastEventId);
+		task->setEventId(++lastEventId);
 	}
 
 	// insert the event id in the list of active events
-	uint32_t eventId = task->getEventId();
-	eventIds.insert(eventId);
-
-	// add the event to the queue
-	eventList.push(task);
+	boost::asio::post(io_context, [this, task]() {
+		auto& timer = eventIdTimerMap[task->getEventId()];
+		timer.reset(new boost::asio::deadline_timer(io_context));
```

Comment:
I don't think there's a reason for the indirection in the mapped type, is there?

Reply:
@djarek at first I didn't think that a task and an entry in the timer map were 1:1, but I just noticed they are and I can cleanup this a bit.